lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
       Pair.value().Index = Pair.index();
 
     return Result;
   }
 
-  static llvm::SmallVector<model::Argument, 8>
-  convertStackArguments(model::QualifiedType StackArgumentTypes,
-                        size_t IndexOffset) {
+  /// Helper used for converting stack argument struct into
+  /// the c-style representation
+  ///
+  /// \param StackArgumentTypes The qualified type of the relevant part of
+  ///        the stack.
+  /// \param IndexOffset The index of the first argument (indicates the register
+  ///        argument count).
+  ///
+  /// \return An ordered list of arguments.
+  static std::optional<llvm::SmallVector<model::Argument, 8>>
+  tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                              size_t IndexOffset) {
+    // Qualifiers here are not allowed since this must point to a simple struct.
     revng_assert(StackArgumentTypes.Qualifiers.empty());
+
     auto *Unqualified = StackArgumentTypes.UnqualifiedType.get();
-    if (not Unqualified)
-      return {};
+    if (!Unqualified) {
+      // If there is no type, it means that the importer responsible for this
+      // function didn't detect any stack arguments and avoided creating
+      // a new empty type.
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
 
     auto *Pointer = llvm::dyn_cast<model::StructType>(Unqualified);
     revng_assert(Pointer != nullptr,
                  "`RawFunctionType::StackArgumentsType` must be a struct");
     const model::StructType &Types = *Pointer;
 
+    // If the struct is empty, there's no arguments to extract from it.
+    if (Types.Fields.empty()) {
+      revng_assert(Types.size() == 0);
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
+
+    // Compute the alignment of the first argument.
+    auto CurrentAlignment = *abi::naturalAlignment(Types.Fields.begin()->Type,
+                                                   ABI);
+    if ((CurrentAlignment & (CurrentAlignment - 1)) == 0) {
+      // Abandon cases where alignment is equal to 0 or is not a power of two.
+      return std::nullopt;
+    }
+
+    // Process each of the fields, converting them into arguments separately.
     llvm::SmallVector<model::Argument, 8> Result;
-    for (const model::StructField &Field : Types.Fields) {
+    for (auto Iterator = Types.Fields.begin();
+         Iterator != std::prev(Types.Fields.end());
+         ++Iterator) {
+      const auto &CurrentArgument = *Iterator;
+      auto MaybeSize = CurrentArgument.Type.size();
+      revng_assert(MaybeSize.has_value() && MaybeSize.value() != 0);
+      uint64_t CurrentSize = paddedSizeOnStack(MaybeSize.value());
+
+      const auto &NextArgument = *std::next(Iterator);
+      auto NextAlignment = *abi::naturalAlignment(NextArgument.Type, ABI);
+      if ((CurrentAlignment & (CurrentAlignment - 1)) == 0) {
+        // Abandon cases where alignment is equal to 0 or is not a power of two.
+        return std::nullopt;
+      }
```

Comment:
(leaving a review to myself so I don't forget to fix this)

This looks like a botched copy-paste. It should check the `NextAlinmentInstead`.

Reply:
This is fixed now.